# <span style="color:red"> Project Maintenance and Correction üõ†Ô∏è </span>

**We are currently in maintenance mode and actively working on improvements. Please bear with us while we make necessary corrections.**


<span style="color:red"> ------------------------------------------- </span>


# LSC-ncRNA : Large Scale Classification of non-coding RNA 

***LSC-ncRNA*** is the implementation of a sequence-based method that relies on the computation and selection of common ncRNA sequence motifs to provide a set of features for effectively and efficiently classifying ncRNA families using a supervised learning approach. 

## General use 

The method is divided into two main steps:  

- 1) [Motifs computation and selection](#motifs_computation_and_selection)  this is the first program to generate a file at the CSV format that contains vector representations of RNA sequences. 

- 2) [Classification](#fit_predict) this is the second program called with the csv file as input to fit the models and to classify new ncRNAs. 

  
# Details & Experimentation Pipeline 

<!-- Requirements --> 

<h3 id="requirements"> :hammer_and_pick: Requirements</h3> 

*   __`C++14`__ 

*   __`Python3 (at least python 3.6)`__ 

*   __`Biopython`__ 

*   __`Datatable`__ 

*   __`Scikit-learn`__ 

*   __`Pandas`__ 

*   __`Numpy`__ 

## A) Dataset preparation   

Download the ncRNA data, for example from Rfam (https://rfam.xfam.org/). In our case we used the following 3 datasets 

### Dataset 1: Rfam 14.1 seed sequences from 3,016 ncRNAs families 

Download the from the ftp server on https://ftp.ebi.ac.uk/pub/databases/Rfam/14.1/Rfam.seed.gz 

Extract Rfam.seed.gz into the file Rfam.seed. 

Command: `gunzip Rfam.seed.gz`

Divide the file Rfam.seed into each family in a separate files, one for each RNA family, such that the file name for each family is the Rfam family name (as Rfxxxxx  such that (xxxxx: is a number). 

You need to use the C++ code for that in (folder), and comiple if necessary:

Command: `g++ main.cpp -o extrcat -std=c++14`

After that just use :

Command: `./extrcat -in path_to//Rfam.seed -out path_to/Rfam_out_files`


Change the sequence identifiers in the fasta file to the family name >Rfamxxxxx , so that sequence classes are provided for subsequent processing. 

Command: xxx 

  

The first dataset is split into 30% Test and 70% Train. 

The train part  is used to do cross validation experimentation. We generate a sample with 600 families, and a sample with 350 families, and each of them is also split into 30% test and 70% train. 

Command: xxx 

  

### Dataset 2: 88 Rfam families with noise 

***Initial dataset*** : Download the data from https://github.com/bioinformatics-sannio/ncrna-deep. This benchmark dataset is used in the method deepncRNA in the paper "Deep learning predicts short non-coding RNA functions from only raw sequence data" ([paper link](https://journals.plos.org/ploscompbiol/article?id=10.1371/journal.pcbi.1008415)). 

  

The dataset includes 306,016 sequences distributed among 88 RFAM families of small ncRNA such that each family contains more than 400 sequences.  Each dataset class was split into three random subsets for train (84\%), validation (8\%), and test (8\%) in the deepncRNA method. We kept the initial partition of the dataset to be able to compare our results with those of deepncRNA. 

  

***Noise dataset*** : For each sequence in the test dataset, we added x\% noise at its start and its end, i.e., x/2\% at each extremity, such that the percentage x\% of noise compared to the length of the original sequence varied between 0\% and 200\% with steps of 25\%. The noise was generated by randomly shuffling the sequence while preserving the nucleotide and dinucleotide frequency of the original sequence. 

  

  

### Dataset 3: Clans from Rfam 14.8 

***Low sequence similarity dataset*** : the  dataset consists of $36$ classes defined as the clans of RFAM 14.8 containing at least $4$ ncRNA families (https://rfam.xfam.org/clan/). This results in $36$ clans containing $4$ to $11$ families per clan and a total of $199$ families for all $36$ clans. For each clan,  $12$ sequences are selected from each family. 

  

***The high sequence similarity dataset*** : the  dataset consists of $199$ classes defined as the $199$ ncRNA families contained in the $36$ clans from the low sequence similarity dataset. 

  

*** List of used clans: *** : *["CL00051", "CL00003", "CL00069", "CL00106", "CL00038", "CL00054", "CL00002", "CL00014", "CL00102", "CL00001", "CL00117", "CL00021", "CL00057", "CL00111", "CL00118", "CL00066", "CL00112", "CL00116", "CL00010", "CL00005", "CL00027", "CL00053", "CL00063", "CL00100", "CL00119", "CL00004", "CL00032", "CL00034", "CL00035", "CL00040", "CL00096", "CL00093", "CL00015", "CL00121", "CL00036", "CL00045"]* 

  

Each class of each dataset was split into two subsets for train (70\%) and test (30\%). 

  

We use the script in ***dataset/clans_dataset.py*** to download and prepare the dataset. 

  

### Dataset Train Test split and dataset statistics: 

  

We use ***constructTrainTestFiles*** to split the dataset into train and test, and to get the dataset statistics. 

- To split the dataset, for each family (fasta file), we split the sequences in  into train and test file. 

- dataset statistics: for a given folder, for each family (in separate file), we get the following information: nb seqs, min seq len, max seq len, average seq len , and save to csv file. 

  

The program can be use like this: 

constructTrainTestFiles -in \<string\> [-out \<string\> -nf \<integer\> -mins \<integer\> -maxs \<integer\> -pt \<integer\> -m \<string\>] 

  

- -in : \<string\> a path directory for fasta files 

- -out : \<string\> a main path directory the output results 

- -nf : \<integer\> number of families 

- -mins : \<integer\>, min number of sequences (default 4) 

- -maxs : \<integer\>, max number of sequences 

- -pt : \<integer\>, percentage of Test sequences, default value is 30. 

- -m : \<string\> user mode, several mode are available:  

    - -m i: (information) get all information as nb seqs, min seq len, max seq len, average seq len , and save to csv file. 

    - -m s: (sample dataset) get n random families that have nb seqs between min and max, and save them to dir_output 

    - -m sttmm : (split Train Test Min Max) for a given nb of families, and min max number of seqs, split into train and test data 

    - -m sttm : (split Train Test Min) consider only min number of seqs, and split all files in input folder into train and test data 

    - -m stt : (split Train Test) for all files in input folder, split into train and test data 

  

The output will be 1) a folder Train with *percentage_nb_seqs_train* sequences, 2) a folder Test  with *100%-percentage_nb_seqs_train* sequences, and 3) a csv file that contains the information. 

  

**note**, be aware of problem of file encoding, and endline. In our case, we have to change to encoding to UTF8 to avoid extra-space when reading a line (see my answer: https://stackoverflow.com/a/73952980/3429103) 

  

  

## <a name="motifs_computation_and_selection"> B) Motifs computation and selection: 

This is the first step in the method which is the computation and the selection of sequence motifs that allow defining a vectorial representation of ncRNAs sequences. The result is a 2d vector (matrix), that contains the number of occurrences for each motif that exist in the ncRNA sequence. The matrix is saved as a single csv file. 

  

The program can be used like this: 

  

./MatrixCmsNStrNbOccs -in \<string\> [-nf \<integer\> -mins \<integer\> -maxs \<integer\> -minl \<integer\> -maxl \<integer\> -d \<integer\> -b \<integer\> -a \<integer\> -g \<integer\> -tn \<string\>] 

  

  

- *-in* : \<string\> a path directory for fasta files 

- *-tn* : \<string\> experiment name, give a specific name to you your experiment, (default "test") 

- *-nf* : \<integer\> number of families, (**default 10**) 

- *-mins* : \<integer\>, min number of sequences, (**default 4**) 

- *-maxs* : \<integer\>, max number of sequences, (**default 1000**) 

- *-minl* : \<integer\>, min length of motif, (**default 2**) 

- *-maxl* : \<integer\>, max length of motif, (**default 10**) 

- *-d* : \<integer\> (0: false, 1 or other: true), delete sub-motifs, (**default 0**) 

- *-b* : \<integer\> beta (between [0 and 100]), (**default 40**) 

- *-a* : \<integer\>, alpha  (-1 no alpha, or 0,1,2,3,..., , (**default -1**) 

- *-g* : \<integer\> ( >=1), gamma, number of occurrences allowed, (**default 1**) 

  

All the parameters between [] are optional. The path to the directory of fasta files is mandatory. The motifs length (minl, maxl), Beta and experiment name parameters are recommended to use. 

  

Example: 

```shell 

nohup ./MatrixCmsNStrNbOccs -in "/data/ibra/Rfam_14.1_dataset/Rfam14.1_Sample_Train_Test/Rfam_600_Train_Test/Train" -minl 2 -maxl 8 -b 50 -g 1 -tn F_600 > out_F_600 & 

``` 

  

The output csv file name is as follows: del_[No/Yes:-d]_nbF_[test_name:-tn]_min_[-minl]_max_[-maxl]_beta_[-b]_alpha_[-a]_nbOccrs_[-g].csv. For exameple, the previous command produce the following name: del_No_nbF_F_600_min_2_max_8_beta_50_alpha_-1_nbOccrs_1.csv 

  

  

## C) Training and Testing experiments: 

  

This is the step of the selection of classification algorithms that allow to achieve the most accurate classification of ncRNA sequences. 

  

## C.1) selection of classification algorithms: 

To choose the best classification algorithm we use the following python program ***Classification/modelstest.py*** as follows: 

`python3 modelstest.py path_motifs.csv`. The code test the following algorithm ['ext','knn','rdf','gnb', 'dt', 'nlp', 'svc'] using 10-fold cross-validation with a split of 70% for train and 30% for validation. 

  

## <a name="fit_predict"> C.2) Experiments with the chosen models. 

  

To launcn the program we use the main python script as follow: 

*python3 Main.py mod "path/2d_matrix.csv" "path_test_folder"*, where: 

- mod is "EXT" (Extra-tree), "RDF" (Random forest), "NLP" (MultiLayer perceptron) and "VOT" (voting model).  

- path/2d_matrix.csv: the path of the csv matrix generated using the previous program. 

- path_test_folder: test folder, that contain a set of fasta files. 

  

Example: 

  

```shell 

python3 Main.py EXT "/data/ibra/del_No_nbF_F_600_min_6_max_7_beta_50_alpha_-1_nbOccrs_1.csv" "/data/ibra/Rfam_14.1_dataset/Rfam14.1_Sample_Train_Test/Rfam_600_Train_Test/Test" > res_EXT_Single_del_No_nbF_Clans_min_6_max_7_beta_50_alpha_-1_nbOccrs_1 

``` 

  

The training and test experiments generate  different scores: 

- Processing time for 1) training and 2) testing. 

- Scores of training `score Train` and prediction `pred Train`. 

- Accuracy, Precision, Recall, and fbeta_score. 

  

  

  

## D) Blast based classification (blastn(av-s)): 

We use Blast and add post-processing of its results to produce classification. We use a script to do the following steps: 

- a) From the Train folder: gather all sequences in all files in fasta file X. 

- b) use blast to create a database on this single file X (that contains all Train seqs) 

- c) Gather all sequences in Test folder in one single file Y. 

- d) use blast to search sequences in Y against the database for X. 

- e) Process blast result best hit to generate the classification. 

  

To lunch the script `blast_classification.py`, we need to give as argument the path for Train and Test folder, and the experiment name as shown in the following example: 

```shell 

nohup python blast_classification.py "/data/chei2402/ibra/test_infernal/Clans_family_train_test/Train/" "/data/chei2402/ibra/test_infernal/Clans_family_train_test/Test/" Clans > res_blast_class_Clans_time_acc & 

``` 

  

  

## D) Infernal based classification: 

  

  

### supplementary sub-folder:  

contains supplementary information and data for our paper. 

  

  

:busts_in_silhouette: __Authors__ 

`ibrahim Chegrane & Nabil Bendjaa & Aida Ouangraoua`, CoBIUS LAB, Department of Computer Science, Faculty of Science, Universit√© de Sherbrooke, Sherbrooke, Canada 

  

> :bulb: If you are using our algorithm in your research, please cite our recent paper: __Upcoming__  

  

> :e-mail: Contact: Aida[dot]Ouangraoua[at]usherbrooke[dot]ca 

 
